crates/next-core/js/src/dev/hmr-client.ts

```diff
@@ -100,116 +102,334 @@ function handleSocketConnected() {
   }
 }
 
-type AggregatedUpdates = {
-  added: Record<ModuleId, HmrUpdateEntry>;
-  modified: Record<ModuleId, HmrUpdateEntry>;
-  deleted: Set<ModuleId>;
-};
-
-// we aggregate all updates until the issues are resolved
-const chunksWithUpdates: Map<ResourceKey, AggregatedUpdates> = new Map();
+// we aggregate all pending updates until the issues are resolved
+const chunkListsWithPendingUpdates: Map<
+  ResourceKey,
+  { update: ChunkListUpdate; resource: ResourceIdentifier }
+> = new Map();
 
 function aggregateUpdates(
   msg: ServerMessage,
-  hasCriticalIssues: boolean
+  aggregate: boolean
 ): ServerMessage {
   const key = resourceKey(msg.resource);
-  const aggregated = chunksWithUpdates.get(key);
-
-  if (msg.type === "issues" && aggregated == null && hasCriticalIssues) {
-    // add an empty record to make sure we don't call `onBuildOk`
-    chunksWithUpdates.set(key, {
-      added: {},
-      modified: {},
-      deleted: new Set(),
-    });
-  }
+  let aggregated = chunkListsWithPendingUpdates.get(key);
 
   if (msg.type === "issues" && aggregated != null) {
-    if (!hasCriticalIssues) {
-      chunksWithUpdates.delete(key);
+    if (!aggregate) {
+      chunkListsWithPendingUpdates.delete(key);
     }
 
     return {
       ...msg,
       type: "partial",
-      instruction: {
-        type: "EcmascriptChunkUpdate",
-        added: aggregated.added,
-        modified: aggregated.modified,
-        deleted: Array.from(aggregated.deleted),
-      },
+      instruction: aggregated.update,
     };
   }
 
   if (msg.type !== "partial") return msg;
 
   if (aggregated == null) {
-    if (hasCriticalIssues) {
-      chunksWithUpdates.set(key, {
-        added: msg.instruction.added,
-        modified: msg.instruction.modified,
-        deleted: new Set(msg.instruction.deleted),
+    if (aggregate) {
+      chunkListsWithPendingUpdates.set(key, {
+        resource: msg.resource,
+        update: msg.instruction,
       });
     }
 
     return msg;
   }
 
-  for (const [moduleId, entry] of Object.entries(msg.instruction.added)) {
-    const removedDeleted = aggregated.deleted.delete(moduleId);
-    if (aggregated.modified[moduleId] != null) {
-      console.error(
-        `impossible state aggregating updates: module "${moduleId}" was added, but previously modified`
-      );
-      location.reload();
+  aggregated = {
+    resource: msg.resource,
+    update: mergeChunkListUpdates(aggregated.update, msg.instruction),
```

Comment:
Since the HMR updates are now very different (from a simple "EcmascriptChunkUpdate" to a "ChunkListUpdate" that can contain simple chunk updates and complex "EcmascriptMergedChunkUpdate"s), I cut the merging logic into smaller functions that each operate at a specific nested-type's level.